ipa-core/src/protocol/basics/reveal.rs

```diff
@@ -80,49 +85,141 @@ pub trait Reveal<C: Context, const N: usize>: Sized {
-                .recv_channel(ctx.role().peer(Direction::Left))
-                .receive(record_id)
-                .await?;
+impl<'a, B, const N: usize> Reveal<DZKPUpgradedSemiHonestContext<'a, B>, N>
+    for Replicated<Boolean, N>
+where
+    B: ShardBinding,
+    Boolean: Vectorizable<N>,
+{
+    type Output = <Boolean as Vectorizable<N>>::Array;
 
-            Ok(Some(share + left + right))
-        }
+    async fn generic_reveal<'fut>(
+        &'fut self,
+        ctx: DZKPUpgradedSemiHonestContext<'a, B>,
+        record_id: RecordId,
+        excluded: Option<Role>,
+    ) -> Result<Option<Self::Output>, Error>
+    where
+        DZKPUpgradedSemiHonestContext<'a, B>: 'fut,
+    {
+        semi_honest_reveal(ctx, record_id, excluded, self).await
     }
 }
 
 /// This implements the malicious reveal protocol over replicated secret sharings.
 /// It works similarly to semi-honest reveal, the key difference is that each helper sends its share
 /// to both helpers (right and left) and upon receiving 2 shares from peers it validates that they
 /// indeed match.
-impl<'a, F: ExtendableField> Reveal<UpgradedMaliciousContext<'a, F>, 1> for MaliciousReplicated<F> {
+pub async fn malicious_reveal<'fut, C, V, const N: usize>(
+    ctx: C,
+    record_id: RecordId,
+    excluded: Option<Role>,
+    share: &'fut Replicated<V, N>,
+) -> Result<Option<<V as Vectorizable<N>>::Array>, Error>
+where
+    C: Context + 'fut,
+    V: SharedValue + Vectorizable<N>,
+{
+    use futures::future::try_join;
+
+    let left = share.left_arr();
+    let right = share.right_arr();
+    let left_sender =
+        ctx.send_channel::<<V as Vectorizable<N>>::Array>(ctx.role().peer(Direction::Left));
+    let left_receiver =
+        ctx.recv_channel::<<V as Vectorizable<N>>::Array>(ctx.role().peer(Direction::Left));
+    let right_sender =
+        ctx.send_channel::<<V as Vectorizable<N>>::Array>(ctx.role().peer(Direction::Right));
+    let right_receiver =
+        ctx.recv_channel::<<V as Vectorizable<N>>::Array>(ctx.role().peer(Direction::Right));
+
+    // Send shares to the left and right helpers, unless excluded.
+    let send_left_fut =
+        MaybeFuture::future_or_ok(Some(ctx.role().peer(Direction::Left)) != excluded, || {
+            left_sender.send(record_id, right)
+        });
+
+    let send_right_fut =
+        MaybeFuture::future_or_ok(Some(ctx.role().peer(Direction::Right)) != excluded, || {
+            right_sender.send(record_id, left)
+        });
+    try_join(send_left_fut, send_right_fut).await?;
+
+    if Some(ctx.role()) == excluded {
+        Ok(None)
+    } else {
+        let (share_from_left, share_from_right) = try_join(
+            left_receiver.receive(record_id),
+            right_receiver.receive(record_id),
+        )
+        .await?;
+
+        tracing::info!("reveal ({:?}): left {left:?} right {right:?} from left {share_from_left:?} from right {share_from_right:?}", ctx.role());
```

Comment:
that will spam the output I think similarly to another `info!` trace above. Lets change it to `trace!` or remove